File: core/os/device/deviceinfo/cc/windows/query.cpp
/*
 * Copyright (C) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../query.h"

#include <windows.h>

namespace {

std::string getOsName(const OSVERSIONINFOEX& version) {
  bool isNTWorkstation = version.wProductType == VER_NT_WORKSTATION;
  int major = version.dwMajorVersion;
  int minor = version.dwMinorVersion;
  if (major == 10 && isNTWorkstation) {
    return "Windows 10";
  } else if (major == 10 && !isNTWorkstation) {
    return "Windows Server 2016 Technical Preview";
  } else if (major == 6 && minor == 3 && isNTWorkstation) {
    return "Windows 8.1";
  } else if (major == 6 && minor == 3 && !isNTWorkstation) {
    return "Windows Server 2012 R2";
  } else if (major == 6 && minor == 2 && isNTWorkstation) {
    return "Windows 8";
  } else if (major == 6 && minor == 2 && !isNTWorkstation) {
    return "Windows Server 2012";
  } else if (major == 6 && minor == 1 && isNTWorkstation) {
    return "Windows 7";
  } else if (major == 6 && minor == 1 && !isNTWorkstation) {
    return "Windows Server 2008 R2";
  } else if (major == 6 && minor == 0 && isNTWorkstation) {
    return "Windows Vista";
  } else if (major == 6 && minor == 0 && !isNTWorkstation) {
    return "Windows Server 2008";
  } else if (major == 5 && minor == 1) {
    return "Windows XP";
  } else if (major == 5 && minor == 0) {
    return "Windows 2000";
  } else {
    return "";
  }
}

device::ABI* abi(device::ABI* abi) {
  abi->set_name("x86_64");
  abi->set_os(device::Windows);
  abi->set_architecture(device::X86_64);
  abi->set_allocated_memory_layout(query::currentMemoryLayout());
  return abi;
}

}  // namespace

namespace query {

bool queryPlatform(PlatformInfo* info, std::string* errorMsg) {
  DWORD size = MAX_COMPUTERNAME_LENGTH + 1;
  WCHAR host_wide[MAX_COMPUTERNAME_LENGTH + 1];
  if (!GetComputerNameW(host_wide, &size)) {
    errorMsg->append("Couldn't get host name: " +
                     std::to_string(GetLastError()));
    return false;
  }
  char hostName[MAX_COMPUTERNAME_LENGTH * 4 + 1];  // Stored as UTF-8
  WideCharToMultiByte(CP_UTF8,                     // CodePage
                      0,                           // dwFlags
                      host_wide,                   // lpWideCharStr
                      -1,                          // cchWideChar
                      hostName,                    // lpMultiByteStr
                      sizeof(hostName),            // cbMultiByte
                      nullptr,                     // lpDefaultChar
                      nullptr                      // lpUsedDefaultChar
  );

  info->name = hostName;
  info->abis.resize(1);
  abi(&info->abis[0]);

  SYSTEM_INFO sysInfo;
  GetSystemInfo(&sysInfo);
  info->numCpuCores = sysInfo.dwNumberOfProcessors;

  info->osKind = device::Windows;
  OSVERSIONINFOEX osVersion;
  osVersion.dwOSVersionInfoSize = sizeof(osVersion);
  GetVersionEx((OSVERSIONINFO*)(&osVersion));
  info->osName = getOsName(osVersion);
  info->osMajor = osVersion.dwMajorVersion;
  info->osMinor = osVersion.dwMinorVersion;
  info->osPoint = osVersion.dwBuildNumber;

  return true;
}

device::ABI* currentABI() { return abi(new device::ABI()); }

device::VulkanProfilingLayers* get_vulkan_profiling_layers() { return nullptr; }

bool hasAtrace() { return false; }

}  // namespace query

File: vulkan_generator/vulkan_parser/api/query.py
# Copyright (C) 2022 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""This module contains the utilities to query information on Vulkan Types"""

from typing import Optional
from vulkan_generator.vulkan_parser.api import types


def try_get_type_or_deducted_type(typename: str, vulkan_types: types.VulkanTypeInfo) -> Optional[types.VulkanType]:
    """This function will return the type object of given typename.
    If type cannot found, then it will look for the type alias with given typename and return the aliased type of it
    If no type alias can be found with given name, then it will return None"""
    if typename in vulkan_types.all_types:
        return vulkan_types.all_types[typename]
    elif typename in vulkan_types.all_aliases:
        return vulkan_types.all_aliases[typename].aliased_type
    else:
        return None


def get_type_or_deducted_type(typename: str, vulkan_types: types.VulkanTypeInfo) -> types.VulkanType:
    """This function will return the type object of given typename.
    If type cannot found, then it will look for the type alias with given typename and return the aliased type of it"""
    if typename in vulkan_types.all_types:
        return vulkan_types.all_types[typename]

    return vulkan_types.all_aliases[typename].aliased_type


def get_struct_or_deducted_struct(typename: str, vulkan_types: types.VulkanTypeInfo) -> types.VulkanStruct:
    """This function will return the type object of given typename.
    If type cannot found, then it will look for the type alias with given typename and return the aliased type of it"""
    if typename in vulkan_types.structs:
        return vulkan_types.structs[typename]

    aliased_struct = vulkan_types.struct_aliases[typename].aliased_type

    if not isinstance(aliased_struct, types.VulkanStruct):
        raise ValueError(f"Unexpected aliased type: {aliased_struct.typename}")

    return aliased_struct


def get_enum_field_or_deducted_field(enum: types.VulkanEnum, field_name: str) -> types.VulkanEnumField:
    """
    This will return the field for the field name.
    If no field cannot be found, then it will search the alias and deduct the field from alias and return it.
    """
    if field_name in enum.fields:
        return enum.fields[field_name]

    return enum.aliases[field_name].aliased_field
